TickSpec/ScenarioRun.fs

```diff
@@ -1,172 +1,173 @@
+    let cons = 
+        t.GetConstructors()
+        |> Seq.tryFind (fun c -> c.GetParameters().Length = table.Header.Length)
+    match cons with
+    | Some c ->
+        // Try to use type's constructor
+        table.Rows |> Array.iteri (fun y row ->
+            let ps = c.GetParameters()
+            let args = 
+                Array.zip ps row
+                |> Array.map (fun (p,s) -> convertString parsers provider p.ParameterType s)
+            let e = Activator.CreateInstance(t, args)
+            ar.SetValue(e, y)
+        )
+    | None -> 
+        // Try to use type's properties
+        let ps = t.GetProperties()
+        table.Rows |> Array.iteri (fun y row ->
+            let e = Activator.CreateInstance(t)
+            for x = 0 to table.Header.Length-1 do
+                let column = table.Header.[x]
+                let p = ps |> Seq.find (fun p -> String.Compare(p.Name, column, StringComparison.InvariantCultureIgnoreCase)=0)           
+                let value = convertString parsers provider p.PropertyType row.[x]
+                p.SetValue(e, value, [||])
+            ar.SetValue(e, y)
+        )
+    ar
+
+/// Invokes method with match values as arguments
+let invokeStep
+        (parsers:IDictionary<Type,MethodInfo>)
+        (provider:IServiceProvider) 
+        (meth:MethodInfo,args:string[],
+         bullets:string[] option,table:Table option,doc:string option) =
+    let meth = meth |> toConcreteMethod
+    let ps = meth.GetParameters()
+    let buildArgs (xs:string[]) =
+        args |> Array.mapi (fun i s ->
+            let p = ps.[i].ParameterType
+            try convertString parsers provider p s
+            with ex ->
+                #if SILVERLIGHT
+                reraise ()
+                #else
+                let name = ps.[i].Name
+                let message = 
+                    sprintf "Failed to convert argument '%s' of target method '%s' from '%s' to type '%s'" 
+                        name meth.Name s p.Name
+                raise <| ArgumentException(message, name, ex)
+                #endif
+        )
+    let args = buildArgs args
+    let tail =
+        match bullets,table,doc with
+        | Some xs,None,None ->
+            let p = ps.[ps.Length-1]
+            let t = p.ParameterType.GetElementType()
+            [|box (toArray parsers provider t xs)|]
+        | None,Some table,None -> 
+            let p = ps.[ps.Length-1].ParameterType
+            if p = typeof<Table> then [|box table|]
+            elif p.IsArray then [|convertTable parsers provider p table|]
+            else failwith "Expecting table argument"
+        | None,None,Some doc -> [|box doc|]
+        | _,_,_ -> [||]
+    let args = Array.append args tail
+    invoke provider meth args
+
+/// Generate scenario execution function
+let generate events parsers (scenario,lines) =
+    fun () ->
+        /// Type instance provider
+        use provider = new ServiceProvider()
```

Comment:
if you made this an optional arg and/or overload it in a different way, perhaps people could feed in a service provider which is instantiated in terms of a DI container (i.e. allowing you to do a 10 line impl which delegates to Autofac. While this should not be the main path and I'm not entirely averse to having a batteries-included experience, we don't want to build a monster either. Having such an extensibility point may allow you to push the InstanceProvider out into the test project rather than relying on it being a part of TickSpec itself.).

I need to think about this aspect a bit more, but my gut sense is that it should be possible to separate out
- the the DI impl side from TickSpec (i.e. default to the boring ServiceProvider unless a specific one is provided and let that be the way toy feed one in
- the DI wiring from the step impls in the Scenario / Steps impl (have not played with it and your example is too small for me to get the full sense of it)